xls/contrib/integrator/ir_integrator_test.cc

```diff
@@ -121,5 +125,343 @@ TEST_F(IntegratorTest, OneSourceFunction) {
+              UnorderedElementsAre(external_1, external_2));
+  EXPECT_FALSE(integration->GetNodesMappedToNode(external_1).ok());
+  EXPECT_FALSE(integration->GetNodesMappedToNode(external_2).ok());
+}
+
+TEST_F(IntegratorTest, MappingTestSetNodeMappingFailureCases) {
+  auto p = CreatePackage();
+  XLS_ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<IntegrationFunction> integration,
+      std::move(IntegrationFunction::MakeIntegrationFunctionWithParamTuples(
+          p.get(), {})));
+  Function& internal_func = *integration->function();
+  Function external_func("external", p.get());
+
+  XLS_ASSERT_OK_AND_ASSIGN(
+      Node * internal_1,
+      internal_func.MakeNodeWithName<Param>(/*loc=*/std::nullopt, "internal_1",
+                                            p->GetBitsType(1)));
+  XLS_ASSERT_OK_AND_ASSIGN(
+      Node * internal_2,
+      internal_func.MakeNodeWithName<Param>(/*loc=*/std::nullopt, "internal_2",
+                                            p->GetBitsType(2)));
+  XLS_ASSERT_OK_AND_ASSIGN(
+      Node * external_1,
+      external_func.MakeNodeWithName<Param>(/*loc=*/std::nullopt, "external_1",
+                                            p->GetBitsType(3)));
+  XLS_ASSERT_OK_AND_ASSIGN(
+      Node * external_2,
+      external_func.MakeNodeWithName<Param>(/*loc=*/std::nullopt, "external_2",
+                                            p->GetBitsType(4)));
+
+  // Mapping = external_1 -> MapsTo -> external_1
+  // Mapping target must be internal.
+  EXPECT_FALSE(integration->SetNodeMapping(external_1, external_1).ok());
+
+  // Mapping = external_1 -> MapsTo -> external_2
+  // Mapping target must be internal.
+  EXPECT_FALSE(integration->SetNodeMapping(external_1, external_2).ok());
+
+  // Mapping = internal_1 -> MapsTo -> external_1
+  // Mapping target must be internal.
+  EXPECT_FALSE(integration->SetNodeMapping(internal_1, external_1).ok());
+
+  // Mapping = internal_1 -> MapsTo -> internal_1
+  // Cannot map to self.
+  EXPECT_FALSE(integration->SetNodeMapping(internal_1, internal_1).ok());
+
+  // Mapping = internal_2 -> MapsTo -> internal_1
+  // Cannot map internal nodes that are not mapping targets.
+  EXPECT_FALSE(integration->SetNodeMapping(internal_2, internal_1).ok());
+}
+
+TEST_F(IntegratorTest, ParamterPacking) {
+  auto p = CreatePackage();
+  FunctionBuilder fb_a("func_a", p.get());
+  fb_a.Param("a1", p->GetBitsType(2));
+  fb_a.Param("a2", p->GetBitsType(4));
+  XLS_ASSERT_OK_AND_ASSIGN(Function * func_a, fb_a.Build());
+
+  FunctionBuilder fb_b("func_b", p.get());
+  fb_b.Param("b1", p->GetBitsType(6));
+  fb_b.Param("b2", p->GetBitsType(8));
+  XLS_ASSERT_OK_AND_ASSIGN(Function * func_b, fb_b.Build());
+
+  XLS_ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<IntegrationFunction> integration,
+      std::move(IntegrationFunction::MakeIntegrationFunctionWithParamTuples(
+          p.get(), {func_a, func_b})));
+
+  auto GetTupleIndexWithNumBits = [&](long int num_bits) {
```

Comment:
JFYI, our linter complains that `long int` is ideally written as `int64` so changed on merge.

Reply:
Sounds good.